crates/bevy_ecs/src/system/query.rs

```diff
@@ -640,6 +679,99 @@ where
+    ///
+    /// If you need to verify the identity of each item returned,
+    /// add [`Entity`] to your [`Query`].
+    ///
+    /// # Example
+    /// ```rust
+    /// # use bevy_ecs::prelude::*;
+    /// #[derive(Component, PartialEq, Debug)]
+    /// struct A(u64);
+    ///
+    /// let mut world = World::new();
+    /// let entity_1 = world.spawn().insert(A(1)).id();
+    /// let entity_2 = world.spawn().insert(A(2)).id();
+    /// let entity_3 = world.spawn().insert(A(3)).id();
+    ///
+    /// let query_state = world.query::<&A>();
+    /// let a_query = Query::from_state(&mut world, &query_state);
+    /// let mut a_iterator = a_query.get_multiple([entity_3, entity_2, entity_1]).map(|i|i.unwrap());
+    /// assert_eq!(*a_iterator.next().unwrap(), A(3));
+    /// assert_eq!(*a_iterator.next().unwrap(), A(2));
+    /// assert_eq!(*a_iterator.next().unwrap(), A(1));
+    /// ```
+    #[inline]
+    pub fn get_multiple<const N: usize>(
+        &'s self,
+        entities: [Entity; N],
+    ) -> impl Iterator<Item = Result<<Q::ReadOnlyFetch as Fetch>::Item, QueryEntityError>> {
+        entities.into_iter().map(|entity| self.get(entity))
+    }
+
+    /// Returns the query results for the provided Array of [`Entity`]s.
+    ///
+    /// These values follow the order of your input Array (if any).
+    /// In case of nonunique entities, a nonexisting entity or a mismatched component,
+    /// a [`QueryEntityError`] is returned instead.
+    ///
+    /// If you need to verify the identity of each item returned,
+    /// add [`Entity`] to your [`Query`].
+    ///
+    /// If you absolutely cannot afford the overhead of verifying uniqueness in this way,
+    /// you can (carefully) call the unsafe [`get_unchecked`](Self::get_unchecked) method repeatedly instead.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use bevy_ecs::prelude::*;
+    /// #[derive(Component, PartialEq, Debug)]
+    /// struct A(u64);
+    ///
+    /// let mut world = World::new();
+    /// let entity_1 = world.spawn().insert(A(1)).id();
+    /// let entity_2 = world.spawn().insert(A(2)).id();
+    /// let entity_3 = world.spawn().insert(A(3)).id();
+    ///
+    /// let query_state = world.query::<&mut A>();
+    /// let mut a_query = Query::from_state(&mut world, &query_state);
+    /// let mut a_iterator = a_query.get_multiple_mut([entity_1, entity_3]).map(|i|i.unwrap());
+    /// let mut a_1 = a_iterator.next().unwrap();
+    /// let mut a_3 = a_iterator.next().unwrap();
+    ///
+    /// *a_1 = A(11);
+    /// *a_3 = A(33);
+    ///
+    /// // Manually drop references so we can access the `World` again
+    /// std::mem::drop(a_iterator);
+    /// std::mem::drop(a_query);
+    ///
+    /// assert_eq!(*world.get::<A>(entity_1).unwrap(), A(11));
+    /// assert_eq!(*world.get::<A>(entity_2).unwrap(), A(2));
+    /// assert_eq!(*world.get::<A>(entity_3).unwrap(), A(33));
+    /// ```
+    #[inline]
+    pub fn get_multiple_mut<const N: usize>(
```

Comment:
Imo this shouldn't use an iterator or hash set. To avoid allocations and hashing, it should do a brute force in place entity collision check. For small numbers of entities (which the const generic basically enforces) this will be much much faster. I expect 99% of use cases to use 2 or 3 entities (one comparison and three comparisons, respectively). And it should return a fixed size array:

```rust
if let Ok([c1, c2, c3]) = query.get_multiple_mut([e1, e2, e3]) {
}
```

Reply:
Quick and dirty impl that ignores error handling:

```rust

    #[inline]
    pub fn get_multiple<const N: usize>(
        &'s self,
        entities: [Entity; N],
    ) -> Result<[<Q::ReadOnlyFetch as Fetch<'w, 's>>::Item; N], QueryEntityError> {
        for i in 0..entities.len() {
            for j in (i + 1)..entities.len() {
                if entities[i] == entities[j] {
                    panic!("cannot call on matching entities")
                }
            }
        }
        Ok(entities.map(|e| unsafe {
            self.state
                .get_unchecked_manual::<Q::ReadOnlyFetch>(
                    self.world,
                    e,
                    self.last_change_tick,
                    self.change_tick,
                )
                .unwrap()
        }))
    }
    ```